tfhe/src/core_crypto/entities/modulus_switched_lwe_ciphertext.rs

```diff
@@ -0,0 +1,286 @@
+use super::lwe_ciphertext::{LweCiphertext, LweCiphertextOwned};
+use crate::core_crypto::commons::ciphertext_modulus::CiphertextModulus;
+use crate::core_crypto::fft_impl::common::modulus_switch;
+use crate::core_crypto::prelude::*;
+
+/// An object to store a ciphertext in little memory
+/// The modulus of the ciphertext is decreased by rounding and the result is stored in a compact way
+/// The uncompacted result can be used as the input of a blind rotation to recover a low noise lwe
+/// ciphertext
+///
+/// ```
+/// use concrete_csprng::seeders::Seed;
+/// use tfhe::core_crypto::prelude::*;
+/// use tfhe::core_crypto::fft_impl::common::modulus_switch;
+/// use tfhe::core_crypto::prelude::modulus_switched_lwe_ciphertext::PackedModulusSwitchedLweCiphertext;
+///
+/// let log_modulus = 12;
+///
+/// let mut secret_generator = SecretRandomGenerator::<ActivatedRandomGenerator>::new(Seed(0));
+///
+/// // Create the LweSecretKey
+/// let lwe_secret_key =
+///     allocate_and_generate_new_binary_lwe_secret_key::<u64, _>(LweDimension(2048), &mut secret_generator);
+/// let ciphertext_modulus = CiphertextModulus::new_native();
+///
+/// let mut seeder = new_seeder();
+/// let seeder = seeder.as_mut();
+///
+/// let mut encryption_generator =
+///     EncryptionRandomGenerator::<ActivatedRandomGenerator>::new(seeder.seed(), seeder);
+///
+///
+/// // Unsecure parameters, do not use them
+/// let lwe = allocate_and_encrypt_new_lwe_ciphertext(
+///     &lwe_secret_key,
+///     Plaintext(0),
+///     Gaussian::from_standard_dev(StandardDev(0.), 0.),
+///     ciphertext_modulus,
+///     &mut encryption_generator,
+/// );
+///
+/// // Can be stored using much less space than the standard lwe ciphertexts
+/// let compressed = PackedModulusSwitchedLweCiphertext::compress(
+///     &lwe.as_view(),
+///     CiphertextModulusLog(log_modulus as usize),
+/// );
+///
+/// let lwe_ms_ed = compressed.extract();
+///
+/// assert_eq!(
+///     modulus_switch(
+///         decrypt_lwe_ciphertext(&lwe_secret_key, &lwe_ms_ed).0,
+///         CiphertextModulusLog(5)
+///     ),
+///     0
+/// );
+/// ```
+pub struct PackedModulusSwitchedLweCiphertext<Scalar: UnsignedTorus> {
+    packed_coeffs: Vec<Scalar>,
+    lwe_dimension: LweDimension,
+    log_modulus: CiphertextModulusLog,
+    uncompressed_ciphertext_modulus: CiphertextModulus<Scalar>,
```

Comment:
also could be good to have accessors for the dimension, log modulus and ciphertext modulus, if we need to check validity during e.g. conformant deserialization

Reply:
I guess we can add them when needed

Reply:
ok